core/src/main/scala/com/microsoft/azure/synapse/ml/causal/SyntheticDiffInDiffEstimator.scala

```diff
@@ -0,0 +1,147 @@
+      .join(unitIdx, preDf(getUnitCol) === unitIdx(getUnitCol), "left_outer")
+      .select(unitIdxCol, timeIdxCol, getTreatmentCol, getPostTreatmentCol, getOutcomeCol)
+      .localCheckpoint(true)
+
+    // fit time weights
+    val (timeWeights, timeIntercept, timeRMSE, lossHistoryTimeWeights) = fitTimeWeights(
+      handleMissingOutcomes(indexedControlDf, timeIdx.count.toInt)(unitIdxCol, timeIdxCol),
+      size
+    )(unitIdxCol, timeIdxCol)
+
+    // fit unit weights
+    val zetaValue = this.get(zeta).getOrElse(calculateRegularization(df))
+    val (unitWeights, unitIntercept, unitRMSE, lossHistoryUnitWeights) = fitUnitWeights(
+      handleMissingOutcomes(indexedPreDf, timeIdx.count.toInt)(unitIdxCol, timeIdxCol),
+      zetaValue,
+      fitIntercept = true,
+      size.swap
+    )(unitIdxCol, timeIdxCol)
+
+    // join weights
+    val Row(t: Long, u: Long) = df.agg(
+      countDistinct(when(postTreatment, col(getTimeCol))),
+      countDistinct(when(treatment, col(getUnitCol)))
+    ).head
+
+    val indexedDf = df.join(timeIdx, df(getTimeCol) === timeIdx(getTimeCol), "left_outer")
+      .join(unitIdx, df(getUnitCol) === unitIdx(getUnitCol), "left_outer")
+
+    val interactionCol = findInteractionCol(indexedDf.columns.toSet)
+    val weightsCol = findWeightsCol(indexedDf.columns.toSet)
+    val didData = indexedDf.select(
+        col(unitIdxCol),
+        col(timeIdxCol),
+        postTreatment.cast(IntegerType).as(getPostTreatmentCol),
+        treatment.cast(IntegerType).as(getTreatmentCol),
+        outcome
+      ).as("l")
+      .join(timeWeights.as("t"), col(s"l.$timeIdxCol") === col("t.i"), "left_outer")
+      .join(unitWeights.as("u"), col(s"l.$unitIdxCol") === col("u.i"), "left_outer")
+      .select(
+        postTreatment,
+        treatment,
+        outcome,
+        (
+          coalesce(col("t.value"), lit(1d / t)) * // time weights
+            coalesce(col("u.value"), lit(1d / u)) + // unit weights
+            lit(getEpsilon) // avoid zero weights
+          ).as(weightsCol),
+        (treatment * postTreatment).as(interactionCol)
+      )
+
+    val linearModel = fitLinearModel(
+      didData,
+      Array(getPostTreatmentCol, getTreatmentCol, interactionCol),
+      fitIntercept = true,
+      Some(weightsCol)
+    )
+
+    val treatmentEffect = linearModel.coefficients(2)
+    val standardError = linearModel.summary.coefficientStandardErrors(2)
+
+    val summary = DiffInDiffSummary(
+      treatmentEffect,
+      standardError,
+      timeWeights = Some(timeWeights),
+      timeIntercept = Some(timeIntercept),
+      timeRMSE = Some(timeRMSE),
+      unitWeights = Some(unitWeights),
+      unitIntercept = Some(unitIntercept),
+      unitRMSE = Some(unitRMSE),
+      zeta = Some(zetaValue),
+      lossHistoryTimeWeights = Some(lossHistoryTimeWeights.toList),
+      lossHistoryUnitWeights = Some(lossHistoryUnitWeights.toList)
+    )
+
+    copyValues(new DiffInDiffModel(this.uid))
```

Comment:
why do you wrap in a copy?

Reply:
In order to copy the shared params. The `DiffInDiffModel` and `SyntheticDiffInDiffEstimator` share these two params: `unitCol` and `timeCol`.